job-server-extras/src/spark.jobserver/python/PythonContextFactory.scala

```diff
@@ -0,0 +1,218 @@
+    * It can't interrogate the JVM type system, so this method is used as an explicit indicator.
+    *
+    * @return the full canonical class name of the context type
+    */
+  def contextType: String
+
+  /**
+    *
+    * @return The entries with which to populate the PYTHONPATH environment variable when
+    *         launching the python subprocess.
+    */
+  def pythonPath: Seq[String]
+
+  /**
+    * Which process to call to execute the Python interpreter, e.g `python`, `python3`
+    *
+    * @return the executable to call
+    */
+  def pythonExecutable: String
+
+  /**
+    * Any mutable actions which need to be taken before the context is used.
+    */
+  def setupTasks(): Unit
+}
+
+trait PythonContextFactory extends SparkContextFactory {
+
+  type J = PythonJobContainer[C]
+
+  override type C <: PythonContextLike
+
+  /**
+    * Loads the job of the given appName, version, and class path, and validates that it is
+    * the right type of job given the current context type.  For example, it may load a JAR
+    * and validate the classpath exists and try to invoke its constructor.
+    */
+  override def loadAndValidateJob(appName: String,
+                                  uploadTime: DateTime,
+                                  classPath: String,
+                                  jobCache: JobCache): J Or LoadingError = {
+    Try(jobCache.getPythonJob(appName, uploadTime, classPath)) match {
+      case Success(jobInfo) => Good(PythonJobContainer(buildJob(jobInfo.eggPath, classPath)))
+      case Failure(ex: Exception) => Bad(JobLoadError(ex))
+      case Failure(ex) => Bad(JobLoadError(new Exception(ex)))
+    }
+  }
+
+  def buildJob(eggPath: String, modulePath:String): PythonJob[C] =
+    PythonJob[C](eggPath, modulePath, py4JImports)
+
+  /**
+    *
+    * @return List of classes which will be imported through the gateway by the Python process,
+    *         using the form `java_import(gateway.jvm, "org.apache.spark.SparkConf")`
+    */
+  def py4JImports: Seq[String]
+}
+
+object PythonContextFactory {
+  val sparkContextImports =
+    Seq(
+      "org.apache.spark.SparkConf",
+      "org.apache.spark.api.java.*",
+      "org.apache.spark.api.python.*",
+      "scala.Tuple2",
+      "org.apache.spark.mllib.api.python.*"
+    )
+
+  val sqlContextImports = sparkContextImports ++ Seq(
+    "org.apache.spark.sql.*"
+  )
+
+  val hiveContextImports = sqlContextImports ++ Seq(
+    "org.apache.spark.sql.hive.*"
+  )
+}
+
+trait DefaultContextLikeImplementations {
+
+  self: PythonContextLike =>
+
+  def config: Config
+
+  lazy val logger = LoggerFactory.getLogger(getClass)
+
+  override lazy val pythonPath: Seq[String] = pythonPaths(config)
+
+  protected def pythonPaths(config: Config): Seq[String] = {
+    val envPaths = sys.env.get("PYTHONPATH").map(_.split(":").toSeq).getOrElse(Seq())
+    val configPaths =
+      if (config.hasPath("python.paths")) {
```

Comment:
We should really use this library: net.ceedubs.ficus.  It would allow us to do `config.as[Option[Seq[String]]("python.paths")`